rslib/src/import_export/package/apkg/import/notes.rs

```diff
@@ -0,0 +1,415 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::{
+    borrow::Cow,
+    collections::{HashMap, HashSet},
+    mem,
+    sync::Arc,
+};
+
+use sha1::Sha1;
+
+use super::{media::MediaUseMap, Context, ProgressFn};
+use crate::{
+    import_export::{
+        package::{media::safe_normalized_file_name, LogNote, NoteLog},
+        ImportProgress,
+    },
+    prelude::*,
+    text::{replace_media_refs, strip_html_preserving_media_filenames, CowMapping},
+};
+
+struct NoteContext<'a> {
+    target_col: &'a mut Collection,
+    usn: Usn,
+    normalize_notes: bool,
+    remapped_notetypes: HashMap<NotetypeId, NotetypeId>,
+    target_guids: HashMap<String, NoteMeta>,
+    target_ids: HashSet<NoteId>,
+    media_map: &'a mut MediaUseMap,
+    imports: NoteImports,
+}
+
+#[derive(Debug, Default)]
+pub(super) struct NoteImports {
+    pub(super) id_map: HashMap<NoteId, NoteId>,
+    /// All notes from the source collection as [Vec]s of their fields, and grouped
+    /// by import result kind.
+    pub(super) log: NoteLog,
+}
+
+impl NoteImports {
+    fn log_new(&mut self, note: Note, source_id: NoteId) {
+        self.id_map.insert(source_id, note.id);
+        self.log.new.push(note.into_log_note());
+    }
+
+    fn log_updated(&mut self, note: Note, source_id: NoteId) {
+        self.id_map.insert(source_id, note.id);
+        self.log.updated.push(note.into_log_note());
+    }
+
+    fn log_duplicate(&mut self, mut note: Note, target_id: NoteId) {
+        self.id_map.insert(note.id, target_id);
+        // id is for looking up note in *target* collection
+        note.id = target_id;
+        self.log.duplicate.push(note.into_log_note());
+    }
+
+    fn log_conflicting(&mut self, note: Note) {
+        self.log.conflicting.push(note.into_log_note());
+    }
+}
+
+impl Note {
+    fn into_log_note(self) -> LogNote {
+        LogNote {
+            id: Some(self.id.into()),
+            fields: self
+                .take_fields()
+                .into_iter()
+                .map(|field| {
+                    strip_html_preserving_media_filenames(&field)
+                        .get_owned()
+                        .unwrap_or(field)
+                })
```

Comment:
Perhaps for now instead of trying to enforce a total size of a record, we could just limit any given stripped field to 80 chars? That should at least avoid issues where users have inserted megabytes of text (copying webpages verbatim for example).